### ISSUE TITLE
All usages of async/await syntax should use "ConfigureAwait"

### ISSUE THREAD

**Issue by craigfowler** [elsa-workflows/elsa-core#426]
Throughout Elsa, because it is shared/reusable code which could be hosted in any one of a number of environments, any place where the `await` keyword is used, it should also use the `.ConfigureAwait(false)` method.

The reasons and problem by not doing this is [described in depth in this article](https://medium.com/bynder-tech/c-why-you-should-use-configureawait-false-in-your-library-code-d7837dce3d7f), but it could be summarised as: _"If you are writing reusable/library code and wish to use `async`/`await`, then always use `.ConfigureAwait(false)` with your awaits.  This will prevent accidental thread-deadlocking when the library is consumed in certain environments."_

```csharp
// Bad example:
await DoTheThingAsync();

// Good example:
await DoTheThingAsync().ConfigureAwait(false);
```

## Consequences/symptom
As eluded to above, in some environments - such as those which have a concept of _"a UI thread"_, not using ConfigureAwait can cause an unwanted deadlock on the main (UI) thread.  That will manifest as the whole application hanging and entering an indefinitely deadlocked state.  Unfortunately it is _not always_ possible to 'fix' this problem from the consuming app.  It is far better to 'fix it' from the library.

**Comment by sfmskywalker** [comment]
I think this is the clearest explanation I've encountered so far, because now I finally understand why I see library authors doing `.ConfigureAwait(false)`.

Makes complete sense to do this.

**Comment by hankovich** [comment]
Sorry, but I'm not buying this concept.
`ConfigureAwait (false)` basically forces the task to answer "I don't care" to the question "where do you want to execute your continuation?" So this causes the task to ignore the provided TaskScheduler, meaning you lose the ability to use the scheduler that the caller wants. Note that custom TaskSchedulers is an extremely powerful feature of .NET ecosystem (you can manually manipulate when, where and how your tasks will be executed)

If `ConfigureAwait(false)` is such a cool option, why isn't it a default then?

I think it makes sense to an the approach [csharpfunctionaextensions follows](https://github.com/vkhorikov/CSharpFunctionalExtensions/blob/master/CSharpFunctionalExtensions/Result/Internal/TaskExtensions.cs#L19), i.e. to allow the caller to decide which value he/she wants to use in the [global config](https://github.com/vkhorikov/CSharpFunctionalExtensions/blob/master/CSharpFunctionalExtensions/Result/Configuration/Configuration.cs#L7) (with a default to false)

What do you think?

**Comment by sfmskywalker** [comment]
I'm thinking that I need to do a bit of research and form my own opinion on this topic ;)

> It's interesting that all examples with pros of ConfigureAwait(false) always contain sync-over-async (i.e. task.Result without awaiting it) :)

I actually assumed this was the defacto use-case for using ConfigureAwait 🤔 
Anway, thanks for raising your concerns, I'm curious to hear any rebuttals.

**Comment by craigfowler** [comment]
I had forgotten about it until you asked but [there's another article with _an incredibly in-depth_ treatment](https://devblogs.microsoft.com/dotnet/configureawait-faq/) of the matter, from the dotnet blog.  I think your question is answered amongst that, _but I hope you like reading_ :laughing:.  My takeaway was quite simply "if it's an environment-neutral library, ConfigureAwait false".

Importantly, that article points out that dotnet library code already uses `ConfigureAwait(false)` just about everywhere.  If you expect that in practice you could use a custom Task Scheduler/Sync Context to run absolutely every task, including ones created via async/await, then actually I don't think it would work anyway, even with official dotnet library code.

The bottom line in that the article seems to be that there's no disadvantage to using `ConfigureAwait(false)` unless you have logic _which absolutely must run in the same thread-context as the caller_.  But if you have such logic, it's not going to be in a shared library, it will be a part of your app.  I guess if the calling app wants tasks (including tasks created by a consumed library) to specifically be added to a Task Scheduler then it is welcome to enqueue them to that scheduler manually, or wrap their calls in a decorator/adapter which does as much.

The idea of making the operation configurable seems sort of OK, but it's different to the standard/recommended practice.  Devs should expect configure await false in library code and should be explicit when they want such tasks to run in their own scheduler.  There doesn't seem to be anything meaningfully gained by allowing otherwise.  Finally, the article does point out (at the end) that this is not an ideal situation, rather it's just _the best they could think of at the time_.  It lists a few proposals about how this might change in future.

**Comment by sfmskywalker** [comment]
Well, consider me schooled. Turns out I knew even less than I already thought I did. 😅 

@craigfowler Thank you for providing that excellent resource! I haven't read everything in detail yet, but after skimming over it (including the referenced GitHub issues at the end) I still agree that it's best if Elsa follows this pattern.

@hankovich I'm curious to hear your thoughts too after reading Stephen's FAQ.
We can of course consider following CSharpFunctionExtension's example of reducing the `ConfigureAwait(false)` noise with an extension. Although I don't see any benefits to allowing the flag to be configurable, given Stephen's explanation on why it's pretty much useless to ever set this to `true`.

**Comment by hankovich** [comment]
Who am I to argue with Mr. Toub? 😊

I'm 99% percent sure you're right and nobody will set it to true, so I'm ok to follow yagni here.

The next question is, how do you force developers to use ConfigureAwait everywhere?
I see two options:
1. Voodoo magic provided by Fody (https://github.com/Fody/ConfigureAwait) that will just automatically add it to all awaits.
2. Play with verbosity of the inspection CA2007.

I guess which option you will choose 😊

**Comment by craigfowler** [comment]
For those who don't use Fody, but who do use code-quality analysis tools such as Sonarqube, [there are rules available](https://rules.sonarsource.com/csharp/RSPEC-3216) which nag you if you don't do it.

In fact I was considering opening an issue suggesting that this project is wired up to Sonarcloud via AppVeyor (which I see you're already using).  It's relatively easy to do and the service is free for FOSS projects.  It's certainly taught me a few things over the years I've been using it, and nipped a few mistakes of my own in the bud too.

**Comment by sfmskywalker** [comment]
Setting up Sonarcloud sounds like a great idea :+1:

**Comment by Rafael94** [comment]
Fody and `ConfigureAwait(false)` is implemented for Elsa 2.0 in #524
